### ISSUE TITLE
Feature: Run salt and capture the result into a job variable.

### ISSUE THREAD

**Issue by zerthimon** [jenkinsci/saltstack-plugin#14]
I think some users (like me) would benefit from the possibility to make salt calls to get an output from salt (e.g: query for a pillar or mine value) into a variable, that can then be used in build steps that follow.

**Comment by mchugh19** [comment]
Great idea. Will do.

**Comment by mchugh19** [comment]
This should do it. Can you test and let me know if it works for you?

https://github.com/jenkinsci/saltstack-plugin/releases/tag/1.7.0-beta3

**Comment by zerthimon** [comment]
Wow, that was fast!
And it works - it sets the SALTBUILDOUPUT var with the output of the command !.
Just a thought: can you make the variable name configurable ?

P.S:
I tried to get a "raw"-formatted json instead of the indented one using out=raw in kwargs, but salt doesn't let me. Says: "ERROR executing 'pillar.get': The following keyword arguments are not valid: out=raw"
Do you know how can I achieve such result ?

**Comment by mchugh19** [comment]
In beta4 I removed the formatting on the envvar. So it should just be the json output from the saltapi. If you ever need it formatted you can always add a shell build step of echo $SALTBUILDOUTPUT | python -m json.tool

**Comment by zerthimon** [comment]
Got an NPE cause of wrong kwarg:
```
FailedConsole Output

Started by user zerthimon
[EnvInject] - Loading node environment variables.
Building in workspace /var/lib/jenkins/jobs/test/workspace
Response on pillar.get deploy_server:project_configs:feeds:static for infra01:
[{"infra01": "ERROR executing 'pillar.get': The following keyword arguments are not valid: out=raw"}]
ERROR: Build step failed with exception
java.lang.NullPointerException
	at com.waytta.SaltAPIBuilder.perform(SaltAPIBuilder.java:396)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
	at hudson.model.Build$BuildExecution.build(Build.java:205)
	at hudson.model.Build$BuildExecution.doRun(Build.java:162)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534)
	at hudson.model.Run.execute(Run.java:1738)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Build step 'Send a message to Salt API' marked build as failure
Notifying upstream projects of job completion
Finished: FAILURE
```

**Comment by zerthimon** [comment]
Queried pillar as before: I see the query result in the log, but the SALTBUILDOUPUT is not defined.

**Comment by mchugh19** [comment]
Gah! My apologies. I'll get that resolved. 

Currently I've got it only saving the envvar, on build success. Perhaps it would be valuable to always capture the results.

**Comment by mchugh19** [comment]
Sorry about that. I forgot to set a default value for the newly added envvar checkbox. This should resolve that problem and will hopefully just work :)

https://github.com/jenkinsci/saltstack-plugin/releases/tag/1.7.0-beta5

**Comment by zerthimon** [comment]
This time the buildout variable is set and is in a raw format.

**Comment by zerthimon** [comment]
Now after this works well, how about a configurable variable name ?

**Comment by mchugh19** [comment]
I'd rather not add the configuration complexity that is the additional text field. But you could support saving the output by adding a followup shell step that re-exports to another name or saves output to a file. Saving to a file would allow for further integration with other jenkins plugins as many will read java properties files.

**Comment by zerthimon** [comment]
Agreed!

**Comment by mchugh19** [comment]
Changes released in 1.7.0
